kernel/aster-nix/src/util/mod.rs

```diff
@@ -1,41 +1,111 @@
 // SPDX-License-Identifier: MPL-2.0
 
+use core::mem;
+
 use aster_frame::vm::VmIo;
 
 use crate::prelude::*;
 pub mod net;
 
-/// copy bytes from user space of current process. The bytes len is the len of dest.
+/// Read bytes into the `dest` buffer
+/// from the user space of the current process.
+/// If successful,
+/// the `dest` buffer is filled with exact `dest.len` bytes.
 pub fn read_bytes_from_user(src: Vaddr, dest: &mut [u8]) -> Result<()> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.read_bytes(src, dest)?)
 }
 
-/// copy val (Plain of Data type) from user space of current process.
+/// Read a value of `Pod` type
+/// from the user space of the current process.
 pub fn read_val_from_user<T: Pod>(src: Vaddr) -> Result<T> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.read_val(src)?)
 }
 
-/// write bytes from user space of current process. The bytes len is the len of src.
+/// Write bytes from the `src` buffer
+/// to the user space of the current process. If successful,
+/// the write length will be equal to `src.len`.
 pub fn write_bytes_to_user(dest: Vaddr, src: &[u8]) -> Result<()> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.write_bytes(dest, src)?)
 }
 
-/// write val (Plain of Data type) to user space of current process.
+/// Write `val` to the user space of the current process.
 pub fn write_val_to_user<T: Pod>(dest: Vaddr, val: &T) -> Result<()> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.write_val(dest, val)?)
 }
 
-/// read a cstring from user, the length of cstring should not exceed max_len(include null byte)
+/// Read a C string from the user space of the current process.
+/// The length of the string should not exceed `max_len`,
+/// including the final `\0` byte.
+///
+/// This implementation is inspired by
+/// the `do_strncpy_from_user` function in Linux kernel.
+/// The original Linux implementation can be found at:
+/// <https://elixir.bootlin.com/linux/v6.0.9/source/lib/strncpy_from_user.c#L28>
 pub fn read_cstring_from_user(addr: Vaddr, max_len: usize) -> Result<CString> {
-    let mut buffer = vec![0u8; max_len];
-    read_bytes_from_user(addr, &mut buffer)?;
-    Ok(CString::from(CStr::from_bytes_until_nul(&buffer)?))
+    let mut buffer: Vec<u8> = Vec::with_capacity(max_len);
+    let mut cur_addr = addr;
+
+    macro_rules! read_one_byte_at_a_time_while {
+        ($cond:expr) => {
+            while $cond {
+                let byte = read_val_from_user::<u8>(cur_addr)?;
+                buffer.push(byte);
+                if byte == 0 {
+                    return Ok(CString::from_vec_with_nul(buffer)?);
```

Comment:
Maybe we could just unwrap `CString::from_vec_with_nul(buffer)` since `byte == 0`. Indeed the function cannot fail. If it fails this is a bad system state. Returning to the user as `E2BIG` when this happens seems not the right way.

I suggest:
```rust
return Ok(CString::from_vec_with_nul(buffer).except("We provided 0 but no 0 is found"));
```

to look as if we care about error handling.